File: 3947_maximum_number_of_items_from_sale_ii/Solution.cs
// LeetCode 3947 - Maximum Number of Items From Sale II
// https://leetcode.com/problems/maximum-number-of-items-from-sale-ii/

using System;
using System.Collections.Generic;

public class Solution {
    public int MaxItems(int[][] items, int budget) {
        int n = items.Length;
        int[] frequency = new int[n + 1];
        int minimumPrice = items[0][1];
        foreach (var item in items) {
            frequency[item[0]]++;
            minimumPrice = Math.Min(minimumPrice, item[1]);
        }
        var batches = new List<(int price, int count)>();
        foreach (var item in items) {
            int gain = 0;
            for (int multiple = item[0]; multiple <= n; multiple += item[0]) gain += frequency[multiple];
            gain--;
            if (gain > 0 && item[1] < 2 * minimumPrice) batches.Add((item[1], gain));
        }
        batches.Sort((a, b) => a.price.CompareTo(b.price));
        long remaining = budget;
        long answer = budget / minimumPrice;
        long boosted = 0;
        foreach (var current in batches) {
            long count = current.count;
            long affordable = remaining / current.price;
            if (affordable < count) count = affordable;
            remaining -= count * current.price;
            boosted += count;
            long total = 2 * boosted + remaining / minimumPrice;
            if (total > answer) answer = total;
            if (count < current.count) break;
        }
        return (int)answer;
    }
}

File: 3947_maximum_number_of_items_from_sale_ii/solution.py
# LeetCode 3947 - Maximum Number of Items From Sale II
# https://leetcode.com/problems/maximum-number-of-items-from-sale-ii/

from typing import List


class Solution:
    def maxItems(self, items: List[List[int]], budget: int) -> int:
        n = len(items)
        frequency = [0] * (n + 1)
        minimum_price = items[0][1]
        for item in items:
            frequency[item[0]] += 1
            minimum_price = min(minimum_price, item[1])
        batches = []
        for item in items:
            gain = 0
            multiple = item[0]
            while multiple <= n:
                gain += frequency[multiple]
                multiple += item[0]
            gain -= 1
            if gain > 0 and item[1] < 2 * minimum_price:
                batches.append([item[1], gain])
        batches.sort(key=lambda a: a[0])
        remaining = budget
        answer = budget // minimum_price
        boosted = 0
        for current in batches:
            count = current[1]
            affordable = remaining // current[0]
            if affordable < count:
                count = affordable
            remaining -= count * current[0]
            boosted += count
            total = 2 * boosted + remaining // minimum_price
            if total > answer:
                answer = total
            if count < current[1]:
                break
        return answer
